### PR TITLE
created Esdl mapper for the pipe and updated the test.

### PR DESCRIPTION
**Pull request by AshrafAlmohagryDeltares**
Please review it.

### CODE REVIEW

File: src/omotes_simulator_core/adapter/transforms/mappers.py [hunk 1]
```diff
@@ -174,9 +176,14 @@ def _convert_assets(self, network: Network) -> List[AssetAbstract]:
             # Esdl Junctions need to be skipped for now, are added later.
             if isinstance(esdl_asset.esdl_asset, esdl_junction):
                 continue
-            py_assets_list.append(EsdlAssetMapper.to_entity(esdl_asset))
-            py_assets_list[-1].add_physical_data(esdl_asset=esdl_asset)
-            network.add_existing_asset(py_assets_list[-1].solver_asset)
+            if isinstance(esdl_asset.esdl_asset, esdl_Pipe):
```
Comment (samvanderzwan) [inline src/omotes_simulator_core/adapter/transforms/mappers.py:179]: This should not be changed in this PR. You can still call the ESDLAssetMapper, since in there you have implemented the logic of selecting to what type it needs to be converted. We can discussed but in different issue if we want to change this. So please reverted these change.

File: unit_test/entities/test_pipe.py [hunk 2]
```diff
@@ -21,38 +21,83 @@
...
+            port_ids=["test1", "test2"],
+            length=self.length,
+            inner_diameter=self.inner_diameter,
+            roughness=self.roughness,
+            alpha_value=self.alpha_value,
+            minor_loss_coefficient=self.minor_loss_coefficient,
+            external_temperature=self.external_temperature,
+            qheat_external=self.qheat_external,
+        )
+        solver_pipe = SolverPipe(name="pipe", _id="pipe_id", length=pipe.length)
```
Comment (samvanderzwan) [inline unit_test/entities/test_pipe.py:55]: Why did you add this? Now you have to acts in your act block. You should only have one act statement. The solver pipe is tested in its own test case

File: unit_test/entities/test_pipe.py [hunk 3]
```diff
@@ -21,38 +21,83 @@
...
         # Assert
         self.assertIsInstance(pipe, Pipe)
         self.assertEqual(pipe.name, "pipe")
         self.assertEqual(pipe.asset_id, "pipe_id")
+        self.assertEqual(solver_pipe.length, self.length)
```
Comment (samvanderzwan) [inline unit_test/entities/test_pipe.py:60]: Why do you test the solver pipe? You should test the pipe. and compare all quantities

File: unit_test/entities/test_pipe.py [hunk 4]
```diff
@@ -21,38 +21,83 @@
...
+            inner_diameter=self.inner_diameter,
+            roughness=self.roughness,
+            alpha_value=self.alpha_value,
+            minor_loss_coefficient=self.minor_loss_coefficient,
+            external_temperature=self.external_temperature,
+            qheat_external=self.qheat_external,
+        )
 
-        # Act
+        solver_pipe = SolverPipe(
```
Comment (samvanderzwan) [inline unit_test/entities/test_pipe.py:78]: This test was OK, so please put it back

File: unit_test/entities/test_pipe.py [hunk 5]
```diff
@@ -64,7 +109,18 @@ def test_pipe_get_property_diameter(self):
...
+            length=self.length,
+            inner_diameter=self.inner_diameter,
+            roughness=self.roughness,
+            alpha_value=self.alpha_value,
+            minor_loss_coefficient=self.minor_loss_coefficient,
+            external_temperature=self.external_temperature,
+            qheat_external=self.qheat_external,
+        )
         esdl_asset_mock = Mock()
         esdl_asset_mock.get_property.return_value = (1.0, False)
```
Comment (samvanderzwan) [inline unit_test/entities/test_pipe.py:125]: This test should be moved to the test of the mapper of the pipe.

File: unit_test/entities/test_pipe.py [hunk 6]
```diff
@@ -87,9 +143,21 @@ def test_pipe_get_heat_transfer_coefficient(self):
...
+            length=self.length,
+            inner_diameter=self.inner_diameter,
+            roughness=self.roughness,
+            alpha_value=self.alpha_value,
+            minor_loss_coefficient=self.minor_loss_coefficient,
+            external_temperature=self.external_temperature,
+            qheat_external=self.qheat_external,
+        )
+
         alpha_value = pipe._get_heat_transfer_coefficient(esdl_asset=esdl_pipe)  # act
```
Comment (samvanderzwan) [inline unit_test/entities/test_pipe.py:161]: This method is moved so this test should be moved.

**Review by samvanderzwan — CHANGES_REQUESTED** [review]
HI Ashraf, I noted several thing we need to change if you want we can discuss, give me a call when you have time.

File: src/omotes_simulator_core/adapter/transforms/esdl_asset_mapper.py [hunk 7]
```diff
@@ -172,3 +185,64 @@ def to_entity(self, esdl_asset: EsdlAssetObject) -> ControllerStorage:
             profile=profile,
         )
         return contr_storage
+
+
+class EsdlAssetPipeMapper(EsdlMapperAbstract):
```
[the remark this answers is not in the document]
Comment (AshrafAlmohagryDeltares) [reply src/omotes_simulator_core/adapter/transforms/esdl_asset_mapper.py:190]: completed.

File: src/omotes_simulator_core/adapter/transforms/esdl_asset_mapper.py [hunk 8]
```diff
@@ -172,3 +185,64 @@ def to_entity(self, esdl_asset: EsdlAssetObject) -> ControllerStorage:
...
+        """Map a Pipe entity to an EsdlAsset."""
+        raise NotImplementedError("EsdlAssetPipeMapper.to_esdl()")
+
+    def to_entity(self, esdl_asset: EsdlAssetObject) -> AssetAbstract:
+        """Method to map an ESDL asset to a pipe entity class.
+
+        :param EsdlAssetObject esdl_asset: Object to be converted to a pipe entity.
+        :return: Pipe object.
+        """
+        # Note: Need to add a check if the returned value is false or not. Some values can be zero.
```
[the remark this answers is not in the document]
Comment (AshrafAlmohagryDeltares) [reply src/omotes_simulator_core/adapter/transforms/esdl_asset_mapper.py:203]: completed.

File: src/omotes_simulator_core/adapter/transforms/esdl_asset_mapper.py [hunk 9]
```diff
@@ -172,3 +185,64 @@ def to_entity(self, esdl_asset: EsdlAssetObject) -> ControllerStorage:
...
+                "external_temperature", PIPE_DEFAULTS.external_temperature
+            )[0],
+            qheat_external=esdl_asset.get_property("qheat_external", PIPE_DEFAULTS.qheat_external)[
+                0
+            ],
+        )
+
+        return pipe_entity
+
+    def _get_heat_transfer_coefficient(self, esdl_asset: EsdlAssetObject) -> float:
```
[the remark this answers is not in the document]
Comment (AshrafAlmohagryDeltares) [reply src/omotes_simulator_core/adapter/transforms/esdl_asset_mapper.py:225]: I added it. completed.

File: src/omotes_simulator_core/adapter/transforms/mappers.py (hunk 1 above)
Comment (AshrafAlmohagryDeltares) [reply src/omotes_simulator_core/adapter/transforms/mappers.py:179]: Completed.

File: unit_test/entities/test_pipe.py (hunk 2 above)
Comment (AshrafAlmohagryDeltares) [reply unit_test/entities/test_pipe.py:55]: completed.

File: unit_test/entities/test_pipe.py (hunk 3 above)
Comment (AshrafAlmohagryDeltares) [reply unit_test/entities/test_pipe.py:60]: completed.

File: unit_test/entities/test_pipe.py (hunk 4 above)
Comment (AshrafAlmohagryDeltares) [reply unit_test/entities/test_pipe.py:78]: completed.

File: unit_test/entities/test_pipe.py (hunk 5 above)
Comment (AshrafAlmohagryDeltares) [reply unit_test/entities/test_pipe.py:125]: I can't find the test of the pipe mapper.

File: src/omotes_simulator_core/entities/assets/esdl_asset_object.py [hunk 10]
```diff
@@ -41,18 +42,30 @@ def __init__(self, asset: esdl.Asset) -> None:
...
+        :param default_value: The default value to return if the property has no value.
         :return: Tuple with the value of the property and a boolean indicating whether the property
-        was found in the esdl_asset.
+                was found in the esdl_asset.
+        If the property is 0, then should it be False or true?
         """
         try:
-            return getattr(self.esdl_asset, esdl_property_name), True
+            value = getattr(self.esdl_asset, esdl_property_name)
+            if value == 0:
```
Comment (vanmeerkerk) [inline src/omotes_simulator_core/entities/assets/esdl_asset_object.py:64]: I don't agree with this change. @samvanderzwan what is your opinion?

File: src/omotes_simulator_core/entities/assets/pipe.py [hunk 11]
```diff
@@ -47,23 +47,42 @@ class Pipe(AssetAbstract):
...
+        minor_loss_coefficient: float,
+        external_temperature: float,
+        qheat_external: float,
+    ):
         """Initialize a Pipe object.
 
         :param str asset_name: The name of the asset.
         :param str asset_id: The unique identifier of the asset.
         :param List[str] port_ids: List of ids of the connected ports.
+        :param float length: The length of the pipe.
```
Comment (vanmeerkerk) [inline src/omotes_simulator_core/entities/assets/pipe.py:68]: For reference add the SI-units in the docstring and remove the docstring from above to below. These variables have changed from non user-defined variables to user-defined variables.

File: src/omotes_simulator_core/entities/assets/pipe.py [hunk 12]
```diff
@@ -47,23 +47,42 @@ class Pipe(AssetAbstract):
...
+        :param float qheat_external: The external heat flow into the pipe.
+        :param float roughness: The wall roughness of the pipe.
+        :param float alpha_value: The alpha value of the pipe.
         """
         super().__init__(asset_name=asset_name, asset_id=asset_id, connected_ports=port_ids)
         # Initialize the default values of the pipe
-        self._minor_loss_coefficient = PIPE_DEFAULTS.minor_loss_coefficient
-        self._external_temperature = PIPE_DEFAULTS.external_temperature
-        self._qheat_external = PIPE_DEFAULTS.qheat_external
+        self._minor_loss_coefficient = minor_loss_coefficient
```
Comment (vanmeerkerk) [inline src/omotes_simulator_core/entities/assets/pipe.py:78]: Change to available variable if it something that the user sets (i.e., change _minor_loss... to minor_loss)

Comment (vanmeerkerk) [reply src/omotes_simulator_core/entities/assets/pipe.py:78]: Do the same for the other properties you now explicitly define.

File: unit_test/entities/test_pipe.py (hunk 5 above)
Comment (vanmeerkerk) [reply unit_test/entities/test_pipe.py:134]: You need to create it yourself.

File: unit_test/entities/test_pipe.py (hunk 4 above)
Comment (vanmeerkerk) [reply unit_test/entities/test_pipe.py:87]: Revert test to original state; you don't need to check the SolverPipe (thats'a different asset)!

**Review by vanmeerkerk — CHANGES_REQUESTED** [review]
Give me a call to discuss the tests you still need to create

File: src/omotes_simulator_core/entities/assets/pipe.py (hunk 11 above)
Comment (AshrafAlmohagryDeltares) [reply src/omotes_simulator_core/entities/assets/pipe.py:68]: Completed.

File: src/omotes_simulator_core/entities/assets/pipe.py (hunk 12 above)
Comment (AshrafAlmohagryDeltares) [reply src/omotes_simulator_core/entities/assets/pipe.py:78]: completed.

File: unit_test/entities/test_pipe.py (hunk 4 above)
Comment (AshrafAlmohagryDeltares) [reply unit_test/entities/test_pipe.py:87]: I removed it.

File: unit_test/entities/test_pipe.py (hunk 5 above)
Comment (AshrafAlmohagryDeltares) [reply unit_test/entities/test_pipe.py:134]: I have added it.

File: src/omotes_simulator_core/adapter/transforms/esdl_asset_mapper.py [hunk 13]
```diff
@@ -32,17 +31,24 @@
 from omotes_simulator_core.entities.assets.controller.controller_storage import ControllerStorage
 from omotes_simulator_core.simulation.mappers.mappers import EsdlMapperAbstract, Entity
 
-CONVERSION_DICT: dict[esdl.EnergyAsset, Type[AssetAbstract]] = {
+
+from omotes_simulator_core.adapter.transforms.esdl_asset_mappers.EsdlAssetPipeMapper import (
+    EsdlAssetPipeMapper,
+)
+
+CONVERSION_DICT: dict[type, Type[AssetAbstract]] = {
```
[the remark this answers is not in the document]
Comment (AshrafAlmohagryDeltares) [reply src/omotes_simulator_core/adapter/transforms/esdl_asset_mapper.py:39]: Eventually, we can have a dictionary for all the mappers and then the other dictionary with the assets can then be removed once we have all the mappers.

File: src/omotes_simulator_core/entities/assets/asset_defaults.py [hunk 14]
```diff
@@ -26,6 +26,7 @@
 DEFAULT_MASS_FLOW_RATE = 1.0  # [kg/s]
 DEFAULT_POWER = 500000.0  # [W]
 DEFAULT_MISSING_VALUE = -9999.99  # [-]
+DEFAULT_ROUGHNESS = 10.0e-4  # [m]
```
[the remark this answers is not in the document]
Comment (AshrafAlmohagryDeltares) [reply src/omotes_simulator_core/entities/assets/asset_defaults.py:29]: I took pvc pipe roughness.  Should I change it?

[the remark this answers is not in the document]
Comment (vanmeerkerk) [reply src/omotes_simulator_core/entities/assets/asset_defaults.py:29]: 10E-4 is equal to 1E-3, so what do you actually want to put as a value? Either way you need to change it to be 1E-4 or 1E-3 depending on the value you want to obtain.

File: src/omotes_simulator_core/entities/assets/pipe.py [hunk 15]
```diff
@@ -15,55 +15,78 @@
...
+        self.minor_loss_coefficient = minor_loss_coefficient
+        self.external_temperature = external_temperature
+        self.qheat_external = qheat_external
         # Define properties of the pipe
-        self.length = PIPE_DEFAULTS.length
-        self.diameter = PIPE_DEFAULTS.diameter
-        self.roughness = PIPE_DEFAULTS.k_value
-        self.alpha_value = PIPE_DEFAULTS.alpha_value
+        self.length = length
+        self.diameter = inner_diameter
```
Comment (vanmeerkerk) [inline src/omotes_simulator_core/entities/assets/pipe.py:87]: One final suggestion, use a more explicit description. You are here actually referring to the inner_diameter so make it known!

File: testdata/test1.esdl [hunk 16]
```diff
@@ -26,7 +26,7 @@
         <port xsi:type="esdl:InPort" connectedTo="935fb733-9f76-4a8d-8899-1ad8689a4b12" id="9c258b9d-3149-4720-8931-f4bef1080ec1" carrier="0bd9cb08-2f69-4e97-8ac8-bd87b07e466a_ret" name="In"/>
         <geometry xsi:type="esdl:Point" CRS="WGS84" lon="4.558639526367188" lat="52.148869383489114"/>
       </asset>
-      <asset xsi:type="esdl:Pipe" related="Pipe1_ret" id="Pipe1" length="6267.0" name="Pipe1" innerDiameter="0.5">
+      <asset xsi:type="esdl:Pipe" related="Pipe1_ret" id="Pipe1" length="6267.0" name="# type: ignore1" innerDiameter="0.5">
```
Comment (vanmeerkerk) [inline testdata/test1.esdl:29]: This seems like an odd change,right? Why would you have a name='# type: ignore1"? Did you perform an automatic change?

File: unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+        esdl_file_path = (
+            Path(__file__).parent / ".." / ".." / ".." / ".." / "testdata" / "test1.esdl"
+        )
+        esdl_file_path = str(esdl_file_path)
+        self.esdl_object = EsdlObject(pyesdl_from_file(esdl_file_path))
+        self.mapper = EsdlAssetPipeMapper()
+
+    def test_to_entity_method(self):
+        """Test for to_entity method."""
+        pipes = self.esdl_object.get_all_assets_of_type("pipe")  # act
```
Comment (vanmeerkerk) [inline unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py:52]: This test is missing the arrange act assert block structure. 

Split all the different 'tasks' in the test, so that it is clear for someone what is what in the test.

File: unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+        self.assertGreater(len(pipes), 0, "No pipes found in the ESDL file.")
+        self.esdl_asset = pipes[0]
+        esdl_asset = pipes[0]
+        pipe_entity = self.mapper.to_entity(esdl_asset)
+        self.assertEqual(pipe_entity.name, esdl_asset.esdl_asset.name)
+        self.assertEqual(pipe_entity.asset_id, esdl_asset.esdl_asset.id)
+        self.assertEqual(pipe_entity.connected_ports, esdl_asset.get_port_ids())  # type: ignore
+        self.assertEqual(
+            pipe_entity.length,
+            esdl_asset.get_property("length", PIPE_DEFAULTS.length)[0],  # type: ignore
```
Comment (vanmeerkerk) [inline unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py:62]: WHy are you checking for PIPE_DEFAULTS.length? If this is actually being set I would suggest to double the value in you self.setUp so that you actually know whether the test was succesfull or not. 

You are checking whether something is and filling in the pipe defaults as a default value; which value is actually set? In other words, will the test fail for the correct reasons?

File: unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+        )
+        self.assertEqual(
+            pipe_entity.diameter, self.mapper._get_diameter(esdl_asset)  # type: ignore
```
Comment (vanmeerkerk) [inline unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py:65]: I would suggest to use the same method for all calls: " esdl_asset.get_property("roughness", PIPE_DEFAULTS.roughness)[0]" or similar. Don't use the mapper methods in this specific case.

File: unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+        """Evaluate failure to retrieve diameter from ESDL asset."""
+        # Arrange
+        esdl_asset_mock = Mock()
+        esdl_asset_mock.get_property.return_value = (1.0, False)
+
+        # Act
+        pass
+
+        # Assert
+        with self.assertRaises(NotImplementedError):
```
Comment (vanmeerkerk) [inline unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py:111]: You should follow the below mentioned structure:


```suggestion
        # Act
        with self.assertRaises(ValueError) as cm:
            self.production_cluster.set_setpoints(setpoints=setpoints)

        # Assert
        self.assertIsInstance(cm.exception, ValueError)
        self.assertEqual(
            cm.exception.args[0],
            f"The setpoints {necessary_setpoints.difference(set(setpoints))} are missing.",
        )
```

File: unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+            EsdlAssetPipeMapper._get_diameter(self, esdl_asset=esdl_asset_mock)
+
+    def test_pipe_get_heat_transfer_coefficient(self):
```
Comment (vanmeerkerk) [inline unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py:114]: I would suggest to create a new 'class', with its own setUp so that you can test the getHeatTransfer coefficient method.

**Review by vanmeerkerk — CHANGES_REQUESTED** [review]
I have some more comments. Just call me so that we can discuss.

File: src/omotes_simulator_core/entities/assets/asset_defaults.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (AshrafAlmohagryDeltares) [reply src/omotes_simulator_core/entities/assets/asset_defaults.py:29]: The reason I left it two significant figures is to follow the format of the other default values set previously (e.g Default_pressure 10.0e5 ). But I have updated it now to 1E-3.

File: src/omotes_simulator_core/entities/assets/pipe.py (hunk 15 above)
Comment (AshrafAlmohagryDeltares) [reply src/omotes_simulator_core/entities/assets/pipe.py:87]: completed.

File: testdata/test1.esdl (hunk 16 above)
Comment (AshrafAlmohagryDeltares) [reply testdata/test1.esdl:29]: This should have been by mistake. I fixed it.

File: unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py (hunk 17 above)
Comment (AshrafAlmohagryDeltares) [reply unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py:52]: completed.

File: unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py (hunk 18 above)
Comment (AshrafAlmohagryDeltares) [reply unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py:62]: I am checking the functionality of the mapper. Specifically, if there is a value in the ESDL input, this test verifies that the value is correctly mapped. If there is no value, the mapper should set it to a default value, and this also needs to be checked. Currently, the test only verifies the mapper’s behavior in the case of a missing input value and the use of a default value. However, with new test data, we can also test if the value is available.

File: unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py (hunk 19 above)
Comment (AshrafAlmohagryDeltares) [reply unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py:65]: Completed.

File: unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py (hunk 21 above)
Comment (AshrafAlmohagryDeltares) [reply unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py:132]: This one we discussed together and said that it can be left like this for now since its also within the pipe mapper.

File: unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py (hunk 18 above)
Comment (AshrafAlmohagryDeltares) [reply unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py:73]: I have added test data file with the roughness included.

**Review by vanmeerkerk — APPROVED** [review]
Approved after discussion

**Review by vanmeerkerk — APPROVED** [review]
Reviewed
